src/NServiceBus.Core/Recoverability/FirstLevelRetries/FirstLevelRetriesHandler.cs

```diff
@@ -0,0 +1,46 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Runtime.ExceptionServices;
+    using System.Threading.Tasks;
+    using Logging;
+    using Pipeline;
+
+    class FirstLevelRetriesHandler
+    {
+        public FirstLevelRetriesHandler(FailureInfoStorage storage, FirstLevelRetryPolicy retryPolicy)
+        {
+            this.storage = storage;
+            this.retryPolicy = retryPolicy;
+        }
+
+        public async Task<bool> HandleMessageFailure(ITransportReceiveContext context, Exception ex)
+        {
+            var messageId = context.Message.MessageId;
+            var pipelineUniqueMessageId = messageId;
+
+            var firstLevelRetries = storage.GetFailureInfoForMessage(pipelineUniqueMessageId).FLRetries;
```

Comment:
that seems to be done by all 3 "handlers", is that something we could pass as a parameter instead to avoid duplicate lookups or do we always need to requery to avoid stale data?